packages/core/src/llm/gemini.ts

```diff
@@ -0,0 +1,381 @@
+    const history = mergedMessages.slice(0, -1);
+    const nextMessage = mergedMessages[mergedMessages.length - 1];
+    const messageContent = GeminiHelper.chatMessageToGemini(nextMessage).parts;
+
+    const client = this.session.gemini.getGenerativeModel(this.metadata);
+
+    const chat = client.startChat({
+      history: history.map(GeminiHelper.chatMessageToGemini),
+    });
+
+    return {
+      chat,
+      messageContent,
+    };
+  }
+
+  protected async nonStreamChat(
+    params: GeminiChatParamsNonStreaming,
+  ): Promise<GeminiChatNonStreamResponse> {
+    const { chat, messageContent } = this.prepareChat(params);
+    const result = await chat.sendMessage(messageContent);
+    const { response } = result;
+    const topCandidate = response.candidates![0];
+    return {
+      raw: response,
+      message: {
+        content: response.text(),
+        role: GeminiHelper.ROLES_FROM_GEMINI[
+          topCandidate.content.role as GeminiMessageRole
+        ],
+      },
+    };
+  }
+
+  protected async *streamChat(
+    params: GeminiChatParamsStreaming,
+  ): GeminiChatStreamResponse {
+    const { chat, messageContent } = this.prepareChat(params);
+    const result = await chat.sendMessageStream(messageContent);
+    return streamConverter(result.stream, (response) => {
+      return {
+        text: response.text(),
+        raw: response,
+      };
+    });
+  }
+
+  chat(params: GeminiChatParamsStreaming): Promise<GeminiChatStreamResponse>;
+  chat(
+    params: GeminiChatParamsNonStreaming,
+  ): Promise<GeminiChatNonStreamResponse>;
+  @wrapLLMEvent
+  async chat(
+    params: GeminiChatParamsStreaming | GeminiChatParamsNonStreaming,
+  ): Promise<GeminiChatStreamResponse | GeminiChatNonStreamResponse> {
+    if (params.stream) return this.streamChat(params);
+    return this.nonStreamChat(params);
+  }
+
+  complete(
+    params: LLMCompletionParamsStreaming,
+  ): Promise<AsyncIterable<CompletionResponse>>;
+  complete(
+    params: LLMCompletionParamsNonStreaming,
+  ): Promise<CompletionResponse>;
+  async complete(
+    params: LLMCompletionParamsStreaming | LLMCompletionParamsNonStreaming,
+  ): Promise<CompletionResponse | AsyncIterable<CompletionResponse>> {
+    const { prompt, stream } = params;
+    const client = this.session.gemini.getGenerativeModel(this.metadata);
+
+    if (stream) {
+      const result = await client.generateContentStream(
+        GeminiHelper.messageContentToGeminiParts(prompt),
+      );
+      return streamConverter(result.stream, (response) => {
+        return {
+          text: response.text(),
+          raw: response,
+        };
+      });
+    }
+
+    const result = await client.generateContent(
+      GeminiHelper.messageContentToGeminiParts(prompt),
+    );
+    return {
+      text: result.response.text(),
+      raw: result.response,
+    };
+  }
+
+  private async getEmbedding(prompt: string): Promise<number[]> {
```

Comment:
Better seperate embedding and LLM model usage in two classes (like the other LLM implementations except Ollama)